Proposal: Modify section 3.2.2.4.7 to allow CA Assisted DNS Validation (#1)
slghtr-says/servercert: docs/BR.md

**Pending Prohibition​​**: The use of a behavior described with this label is highly discouraged, as it is planned to be deprecated and will likely be designated as MUST NOT in the future.

**Prefixed Canonical Authorization Domain Name**: The domain contained within the RDATA value of a CNAME record from an Authorization Domain Name that is prefixed with a Domain Label that begins with an underscore character.

> 

Reviewer: That's good, but this should be clarified since it can be read as either:

"The domain [contained within the RDATA value of a CNAME record from an Authorization Domain Name that is prefixed with a Domain Label that begins with an underscore character]"
or:
"[The domain that is prefixed with a Domain Label that begins with an underscore character] that is contained within the RDATA value of a CNAME record from an Authorization Domain Name"

> 

Reviewer: I think my confusion stems from the fact that we're using a term that begins with "prefixed" to refer to an FQDN that is not [typically] prefixed.

I'd suggest:
**Canonical Authorization Domain Name**: The domain name found within the RDATA value of a CNAME record located at an FQDN composed of an Authorization Domain Name that is prefixed with a Domain Label that begins with an underscore character.

> 

Author: Updated with suggested text

##### 3.2.2.4.7 DNS Change

> Confirming the Applicant's control over the FQDN by confirming the presence of a Random Value or Request Token for either in a DNS CNAME, TXT or CAA record for 1) an Authorization Domain Name; or 2) an Authorization Domain Name that is prefixed with a Domain Label that begins with an underscore character; or 3) a Prefixed Canonical Authorization Domain Name.

Reviewer 2: ```suggestion
Confirming the Applicant's control over the FQDN by confirming the presence of a Random Value or Request Token either in a DNS CNAME, TXT or CAA record for 1) an Authorization Domain Name; or 2) an Authorization Domain Name that is prefixed with a Domain Label that begins with an underscore character; or 3) a Prefixed Canonical Authorization Domain Name.
```

If a Random Value is used, the CA SHALL provide a Random Value unique to the Certificate request and SHALL not use the Random Value after

  i. 30 days or
  ii. if the Applicant submitted the Certificate request, the time frame permitted for reuse of validated information relevant to the Certificate (such as in [Section 4.2.1](#421-performing-identification-and-authentication-functions) of these Guidelines or Section 11.14.3 of the EV Guidelines).

> CAs MAY operate domains for the purpose of assisting Applicants with this validation, and MAY instruct Applicants to add a CNAME record containing a Canonical Authorization Domain Name controlled by the CA. If the CA does so, the CA SHALL 

Author: Looking for feedback on using "Applicant" rather than "Account" in this line. See prior thread for context: https://github.com/slghtr-says/servercert/pull/1#discussion_r1501247583

Reviewer 3: I prefer the use of "Applicant". The BRs should describe the goal (i.e., separation of data/validations by Applicant) as opposed to the explicit mechanism (i.e., account segregation on the CA system).

Using the term "Applicant" describes the goal, whereas "Account" describes the mechanism. Additionally, the term "Account" does not currently exist, so it would have to be created for just this case. And we would likely have to amend section 4.2.1 to use this term instead of "Applicant" when speaking to the requirements/allowances for validation data reuse.

Author: Will keep the term "Applicant" and resolve this comment thread.

> CAs MAY operate domains for the purpose of assisting Applicants with this validation, and MAY instruct Applicants to add a CNAME record containing a Prefixed Canonical Authorization Domain Name. If the CA does so, the CA SHALL 

Reviewer 2: Is it clear that this overrules the time and reuse constraints that are put on Random Value and Request Token values? 

Also should this state "containing a Prefixed Canonical Authorization Domain Name **pointing to a CA-owned domain name**"?

Author: Does this make it any clearer?
```suggestion
CAs MAY operate domains for the purpose of assisting Applicants with this validation, and MAY instruct Applicants to add a CNAME record containing a Prefixed Canonical Authorization Domain Name operated by the CA. If the CA does so, the CA SHALL 
```

Reviewer 2: I like it!

Author: **Rationale**: "Control" has a specific meaning in the BRs 

> Control: “Control” (and its correlative meanings, “controlled by” and “under common control with”)
means possession, directly or indirectly, of the power to: (1) direct the management, personnel,
finances, or plans of such entity; (2) control the election of a majority of the directors ; or (3) vote
that portion of voting shares required for “control” under the law of the entity’s Jurisdiction of
Incorporation or Registration but in no case less than 10%

Author: > Is it clear that this overrules the time and reuse constraints that are put on Random Value and Request Token values?

That is not included in the scope of this incremental change. While I understand the value of the Random Value and Request Token are limited in this paradigm modifying them would make this 3rd "variation" of 3.2.4.4.7 meaningfully different from the first 2 variations in practice. I want to avoid that here and push that to the follow-on discussions around a new automation based validation method.

>   i. ensure that each Prefixed Canonical Authorization Domain Name is used for a unique Applicant, and not shared across multiple Applicants; and 

Reviewer 2: Should this say Applicant, or Account?
There's a meaningful difference between Applicant-binding, and Account-binding. I've been under the assumption that we're going to perform Account-binding for this

Author: The mitigation we landed on in the Threat-Model was "Applicant Unique CNAMEs". That said, we may have used the terms applicant and account interchangeably at times. 

The reasons I'm in favor of Applicant is 1) the noun "Account" is not defined or used in the BR's today and 2) if we were to define it, I imagine it would be similar to how "Applicant" is defined. 

> Applicant: The natural person or Legal Entity that applies for (or seeks renewal of) a Certificate.
Once the Certificate is issued, the Applicant is referred to as the Subscriber. For Certificates issued to
devices, the Applicant is the entity that controls or operates the device named in the Certificate, even
if the device is sending the actual certificate request

Based on that definition, I believe ensuring that the CA is generating unique PeCAns (Prefixed Canonical Authorization Domain Names) for each Applicant achieves the desired outcome. 

Does that make sense?

Reviewer 2: @slghtr-says I'm not sure yet on this one. Specially when we also consider rfc8657, which is much more about account binding than applicant binding. 

I expect that for CAs, account binding is something that provides something more meaningful than applicant binding. Account binding means the CA can tie this to a specific account, and thus connected to certain credentails, whereas CAs may not have the same options for Applicants. 

If that means that we need to get Account defined, I'm all for that :) I'll give that some thought

Author: Leaving this thread open for now

>   ii. complete the validation within 8 hours; and

Author: Looking for feedback on 8 hours as a time limit. See conversation from previous iteration for more context: https://github.com/slghtr-says/servercert/pull/1#discussion_r1478572500

Reviewer 3: I like 8 hours as it's a sufficiently small period of time where the security goal of not relying on "old" records is achieved while providing alignment with the CAA "8-hour minimum TTL".

Author: Thanks @CBonnell. Will resolve this thread.

>   ii. complete the validation within the TTL of the CNAME record, or 8 hours, whichever is lesser; and

Author: ```suggestion
  ii. complete the validation within 8 hours.; and
```
The previous language attempts to set a maximum bound on the time the CA can rely on the results of DNS lookup for the purposes of domain validation. Setting the lower bound to the TTL however does not account for practically invalid scenarios such as a TTL of 0 or  less than 30 seconds. Well operated CAs may not be able to meet such strict time constraints. To simplify the constraint, I suggest removing the mention of TTLs and setting the upper bound for the usage of the DNS query results to 8 hours.

Reviewer 2: I worry that 8 hours by default is too much. If the record points towards the CA, the CA could easily re-check if a certain amount of time has expired. Maybe we should limit this down to an hour, otherwise the CA needs to re-check if the record still exists?

Author: **Rationale behind the 8 hours:** The time we are measuring if from when the DNS query is performed until the results are analyzed and the domain is considered "VALID" by the DCV system. I suspect this practice typically occurs in the order of seconds or minutes under normal operating conditions. I don't suspect that CAs will have difficulty meeting an hour (in the event of a technical issue) or performing another lookup if required. That said, 8 hours is the smallest time constraint referenced in the BRs (OCSP & CAA). CAA record checks are a great parallel for this given that CAA DNS look-up have an validity floor of 8 hours. 

Since the BRs do not currently have a constraint at all on this particular practice I think staying in line with existing time constraints is reasonable. Finally, I don't suspect there is a significant increase in risk if a CA is does a DNS look-up, goes down for 6 hours and then continues the validation when the systems come back up.

Author: Removed the TTL constraint. Will open new thread about the 8 hours.

>   iii. restrict the DNS record types in the CA operated DNS zone to those required for confirming the Applicant’s control over the FQDN.

Reviewer 2: I'm trying to find the meaning being. This does show that, for the purpose of delegating DCV to the CA, the CA must use a different domain name, for which no other records are hosted in the DNS zone?

Author: That is the intent but I am open for suggestions on how to make this clearer or more helpful.

Reviewer 3: This might be hard to nail down into something clear. Top of mind is that this language seemingly prohibits the population of DNSSEC-specific record types, as they are not currently "required". However, I don't think such a prohibition on the use of DNSSEC is desirable.

Perhaps we can merely prohibit the inclusion of A and AAAA records?

  i. 30 days or
  ii. if the Applicant submitted the Certificate request, the time frame permitted for reuse of validated information relevant to the Certificate (such as in [Section 4.2.1](#421-performing-identification-and-authentication-functions) of these Guidelines or Section 11.14.3 of the EV Guidelines).

> CAs MAY operate domains for the purpose of assisting customers with this validation, and MAY instruct customers to add a CNAME redirect from an Authorization Domain Name to such a domain. If the CA does so, the CA SHALL 

Reviewer 4: ```suggestion
CAs MAY operate Domain Names for the purpose of assisting Applicants with this validation, and MAY instruct Applicants to add a CNAME record from an Authorization Domain Name to such a Domain Name operated by the CA. If the CA does so, the CA SHALL 
```

>   i. ensure that each domain name is used for a unique Applicant, and not shared across multiple Applicants; and 

Reviewer 4: ```suggestion
  i. ensure that each Domain Name operated by the CA is used for a unique Applicant, and not shared across multiple Applicants; and 
```
